clr/commands.py

```diff
@@ -229,10 +329,40 @@ def cmd_completion1(self, query=''):
             results.extend(f'{k}:' for k in NAMESPACE_KEYS)
         else:
             namespace_key, _ = query.split(':', 1)
-            results.extend(f'{namespace_key}:{c} ' for c in self.cache.get(namespace_key).commands)
+            namespace = self.cache.get(namespace_key)
+            results.extend(f'{namespace_key}:{c} ' for c in namespace.commands)
 
         print('\n'.join(r for r in results if r.startswith(query)), end='')
 
+    def cmd_completion_arg(self, command_name, partial='', bools_only=False):
+        """Completion results for arguments.
+
+        Optionally only prints out the boolean flags."""
+
+        namespace_key, command_name = resolve_command(command_name, cache=self.cache)
+        namespace = self.cache.get(namespace_key)
+        parser = namespace.argument_parser(command_name)
+        # argparse doesn't officially support any introspection. Reach into the
+        # guts and pull out the bits we need. This isn't perfect, but their
+        # internal api has been stable for years and the worst thing that can
+        # break is shell completion. Existing off the shelf solutions like the
+        # argcomplete package work by monkey-patching argparse before you build
+        # your parsers which is potentially even more brittle.
+
+        options = []
+        for action in parser._actions:
+            if bools_only and not isinstance(action, argparse._StoreConstAction):
+                continue
+            options.extend(action.option_strings)
+
+        # partial is prepended with a space to stop argparse from parsing it
+        partial = partial.strip()
+        print('\n'.join(f'{o} ' for o in options if o.startswith(partial)), end='')
+
+    def cmd_argtest(self, a, b, c=4, d=None, e=False):
+        """For testing arg parsing."""
```

Comment:
Is this method meant to be included, or was it just for local testing while you were building this?

Reply:
Removed it and added it enough times I was just going to keep it in here unless there is an objection. Was thinking of using it for unit tests when I get some time to get back to this.

Reply:
Looks like a really good sanity check to make sure the downstream parser works. Might be worth a second version with:
```def cmd_argtest2(self, a, b, *, c=4, d=None, e=False):```